storybookjs/storybook
# Core: Add Babel mode v7

Issue: #14425 #2758 #12745 #15502 #12292

This PR overhauls Storybook's babel handling in preparation for v7.0.

Before:

- Storybook provides a baseline config
- User config is inconsistently respected (sometimes respected and sometimes not!)
- Addons and frameworks add incremental modifications to config
- Users recommended to override `babel` in `main.js`
- NO debugging methodology

After (when `features.babelModeV7` flag is activated):

- Storybook provides NO baseline config
- Users can ask Storybook to generate a baseline config file (`.babelrc.json`)
- User config is respected
- Addons and frameworks add incremental modifications to config
- Users recommended to edit their project babel config, can worst-case override it in `main.js webpackFinal`
- Debugging methodology in docs

## What I did

- [x] Add `features.babelModeV7`
- [x] Update webpack4/5 builder/manager changes
	- DON'T load storybook babel defaults in the new mode
	- DO configure babel loader to look for the project babelrc/babel config
- [x] Fix web-components babel handling for consistency
- [x] Automatically generate babelrc on empty project `sb init`
- [x] CLI command for force-generating babelrc
- [x] Open question: do incremental modifications to babel config actually work??
- [x] Migration instructions
- [x] Documentation w/ debug instructions

## Potential improvements

- [ ] Fix CLI dependency structure
- [ ] Automatically set V7 feature flag on empty projects

## How to test


The easiest way to test this is to modify the `.babelrc` of `examples/react-ts`, which has `babelModeV7` enabled. You can:

- [ ] Manually edit the babelrc
- [ ] Remove the babelrc and generate one using `../../lib/cli/bin/index.js`
- [ ] De-activate the `babelModeV7` flag
- [ ] Activate the `babelModeV7` flag in `examples/cra-kitchen-sink`

You can also do all of this with `BABEL_SHOW_CONFIG_FOR` enabled to see what's going on under the hood. But you can't actually run the Storybooks this way -- see the documentation.

If your answer is yes to any of these, please make sure to include it in your PR.
